File: Programmers/Lv1/키패드 누르기/iamneveralone/키패드 누르기.cpp
#include <string>
#include <vector>

using namespace std;

struct Pos{
    int x, y;  
};

Pos cur_left, cur_right; // 현재 왼손, 오른손 엄지손가락 위치

string solution(vector<int> numbers, string hand) {
    string ans = "";
    
    cur_left = {3, 0}, cur_right = {3, 2};
    
    for (int i = 0; i < numbers.size(); i++){
        int num = numbers[i]; // 현재 누를 숫자
        
        if (num == 1 || num == 4 || num == 7){ // 1, 4, 7
            ans += 'L';
            cur_left = {num / 3, 0};
        } 
        else if (num == 3 || num == 6 || num == 9){ // 3, 6, 9
            ans += 'R';
            cur_right = {num / 3 - 1, 2};
        }
        else { // 2, 5, 8, 0
            Pos num_pos;
            if (num == 0) num_pos = {3, 1};
            else num_pos = {num / 3, 1};
            
            int left_dist = abs(num_pos.x - cur_left.x) + abs(num_pos.y - cur_left.y); // 왼손 엄지손가락과의 거리
            int right_dist = abs(num_pos.x - cur_right.x) + abs(num_pos.y - cur_right.y); // 왼손 엄지손가락과의 거리
            
            if (left_dist < right_dist){
                cur_left = {num_pos.x, num_pos.y};
                ans += "L";
            }
            else if (left_dist > right_dist){
                cur_right = {num_pos.x, num_pos.y};
                ans += "R";
            } 
            else { // 거리가 같다면
                if (hand == "left") {
                    cur_left = {num_pos.x, num_pos.y};
                    ans += "L";
                }
                else if (hand == "right"){
                    cur_right = {num_pos.x, num_pos.y};
                    ans += "R";
                } 
            }
        }
    }
    return ans;
}

File: Programmers/Lv1/키패드 누르기/jiwon2724/키패드 누르기.kt
class Solution {
    var left = intArrayOf(1, 4, 7, 10)
    var right = intArrayOf(3, 6, 9, 12)
    var currentPosition = intArrayOf(10, 12)

    val stringBuilder = StringBuilder()

    fun solution(numbers: IntArray, hand: String): String {
        numbers.forEach {
            when (it) {
                in left -> checkLeft(it)
                in right -> checkRight(it)
                else -> checkHand(hand, it)
            }
        }

        return stringBuilder.toString()
    }

    fun checkLeft(n: Int) {
        currentPosition[0] = n
        stringBuilder.append("L")
    }

    fun checkRight(n: Int) {
        currentPosition[1] = n
        stringBuilder.append("R")
    }

    fun checkHand(hand: String, num: Int) {
        val n = if (num == 0) 11 else num
        val leftDistance = abs(((currentPosition[0] - n) / 3) + ((currentPosition[0] - n) % 3))
        val rightDistance = abs(((currentPosition[1] - n) / 3) + ((currentPosition[1] - n) % 3))

        if (leftDistance > rightDistance) {
            checkRight(n)
        } else if (leftDistance < rightDistance) {
            checkLeft(n)
        } else {
            if (hand == "left") checkLeft(n) else checkRight(n)
        }
    }
}